### ISSUE TITLE
Matrix.invert() method without allocation

### ISSUE THREAD

**Issue by Extrawurst** [Dav1dde/gl3n#50]
currently the mixin AFAIK creates a dynamic array even though the matrix already contains one, just assignments would be enough

**Comment by Dav1dde** [comment]
I agree, but if you look at https://github.com/Dav1dde/gl3n/blob/master/gl3n/linalg.d#L1127-L1158 - It's simply not feasable to do. You would need to store half of the content in local variables and access the other half on the fly (matrix[x][y]). That just generates a huge blob of unreadable code. My hope is/was that the generator somehow can optimize that to a static array.

**Comment by Extrawurst** [comment]
i am not sure about this, i always thought a multidimensional array literal is constructed using gc allocations. i could be wrong.

**Comment by Dav1dde** [comment]
Tbh, I have no idea, I was simply assuming the optimizer has gotten good enough ... Then again, it's D/dmd. But we wont know for sure until someone runs it through a debugger/looks at the asm. 

If you wanna go for it and create the ugly code required for replacing it, I will merge it. But please add unittests beforehand (the code required will be really ugly, I assume, easy to make mistakes).

**Comment by Extrawurst** [comment]
a simple test with array literals on http://d.godbolt.org/ reveil that these create a call to _d_arrayLiteralTX that actually uses a gc_allocation:
https://github.com/D-Programming-Language/druntime/blob/e47a00bff935c3f079bb567a6ec97663ba384487/src/rt/lifetime.d#L2135

so at least we now know that literals create heap traffic :(

**Comment by Dav1dde** [comment]
Yeah that sucks ... if even gcc doesn't optimize it. 
(To answer your question from IRC this morning, yes I am dav1d on freenode)
